### PR TITLE
Release pre-processed batch requests out of the batch lock

### PR DESCRIPTION
**Pull request by yuliasherman**
* **Problem Overview**  
   The locking order of different mutexes has to be preserved to avoid dead-locks. An attempt to fix such an ordering issue 
   resulted in a crash. Introducing an alternative issue resolution.
* **Testing Done**  
  CI

### CODE REVIEW

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 1]
```diff
@@ -192,14 +193,17 @@ void RequestsBatch::releaseReqsAndSendBatchedReplyIfCompleted(PreProcessReplyMsg
     batchCid = batchCid_;
     batchSize = batchSize_;
     // The last batch request has pre-processed => send batched reply message
-    lock.unlock();
-    for (auto const &replyMsg : repliesList_) {
-      replyMsgsSize += replyMsg->size();
-      preProcessor_.releaseClientPreProcessRequestSafe(clientId_, replyMsg->reqOffsetInBatch(), COMPLETE);
+    for (auto const &reply : repliesList_) {
```
Comment (eyalrund) [inline bftengine/src/preprocessor/PreProcessor.cpp:196]: this is only reverting to the previous solution, right?

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 2]
```diff
@@ -177,6 +177,7 @@ void RequestsBatch::releaseReqsAndSendBatchedReplyIfCompleted(PreProcessReplyMsg
   string batchCid;
   atomic_uint32_t batchSize = 0;
   PreProcessBatchReplyMsgSharedPtr batchReplyMsg;
+  set<uint16_t> reqOffsetsInBatch;
```
Comment (ukg19) [inline bftengine/src/preprocessor/PreProcessor.cpp:180]: Nit: Can we use another container which has better insertion time complexity ,here e.g. a vector? which gives amortized constant time insertion complexity while set will give logarithmic to the size of set time complexity?

**Review by ukg19 — APPROVED** [review]

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 2 above)
Comment (yuliasherman) [reply bftengine/src/preprocessor/PreProcessor.cpp:180]: I can change that, but for a few tens of requests, it really does not matter.

**Review by eyalrund — APPROVED** [review]

**Comment by eyalrund** [comment]
we should create a private build and give ST to verify. Also run maesto on this fix

**Pull request merged**